File: Year 3/Parallel and Distributed Programming/Convolution Matrix/ConvolutionMatrix C++/ConvolutionMatrix C++ Parallel/Source Files/Singleton/Convolution.cpp
#include "../../Header Files/Singleton/Convolution.h"
Convolution* Convolution::instance = nullptr;

Convolution* Convolution::getInstance() {
	if (instance == nullptr) {
		instance = new Convolution();
	}
	return instance;
}

int Convolution::applyConvolution(const vector<vector<int>>& inputMatrix,
	const vector<vector<int>>& convolutionMatrix,
	int i, int j) {
	int result = 0;
	int n = inputMatrix.size();
	int m = inputMatrix[0].size();
	int k = convolutionMatrix.size();

	for (int p = 0; p < k; ++p) {
		for (int q = 0; q < k; ++q) {
			int x = min(max(i - k / 2 + p, 0), n - 1);
			int y = min(max(j - k / 2 + q, 0), m - 1);
			result += inputMatrix[x][y] * convolutionMatrix[p][q];
		}
	}

	return result;
}

File: Year 3/Parallel and Distributed Programming/Laboratory 2/Java/src/Singleton/Convolution.java
package Singleton;

public class Convolution {
    private static Convolution instance = null;
    private Convolution() {
    }
    public static Convolution getInstance() {
        if (instance == null) {
            instance = new Convolution();
        }
        return instance;
    }
    public int applyConvolution(int[][] inputMatrix, int[][] convolutionMatrix, int j) {
        int result = 0;
        int n = inputMatrix.length;
        int m = inputMatrix[0].length;
        int k = convolutionMatrix.length;
        for (int p = 0; p < k; p++) {
            for (int q = 0; q < k; q++) {
                int x = 1 - k/2 + p;
                int y;
                if (j == 0)
                    y = Math.max(j - k/2 + q, 0);
                else if (j == m - 1)
                    y = Math.min(j - k/2 + q, m - 1);
                else
                    y = j - k/2 + q;
                result += inputMatrix[x][y] * convolutionMatrix[p][q];
            }
        }
        return result;
    }
}
